In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np

In [ ]:
base_dir = 'drive/MyDrive/WineQT/'

In [ ]:
inputs,targets = pd.read_csv(base_dir+'wineqt_2_inputs.csv'),pd.read_csv(base_dir+'wineqt_2_targets.csv')

In [ ]:
targets = targets.drop(['Unnamed: 0'],axis=1)

In [ ]:
targets_dummies = pd.get_dummies(targets)

In [ ]:
targets_dummies

quality_bad  quality_good
0               1             0
1               1             0
2               1             0
3               0             1
4               1             0
...           ...           ...
1039            1             0
1040            1             0
1041            0             1
1042            1             0
1043            1             0

[1044 rows x 2 columns]

In [ ]:
inputs_train,inputs_test,targets_train,targets_test = train_test_split(inputs,targets_dummies,test_size=0.15,random_state=10)

In [ ]:
scaler = StandardScaler()
scaler.fit(inputs)

StandardScaler()

In [ ]:
inputs_scaled_train = scaler.transform(inputs_train)
inputs_scaled_test = scaler.transform(inputs_test)

## Logistic Regression

>> Model Making


In [ ]:
model = LogisticRegression()
model.fit(inputs_scaled_train,targets_train.iloc[:,1])

LogisticRegression()

>> Train accuracy/score

In [ ]:
f'{model.score(inputs_scaled_train,targets_train.iloc[:,1])*100}%'

'76.09921082299887%'

>> Test accuracy/score

In [ ]:
f'{model.score(inputs_scaled_test,targets_test.iloc[:,1])*100}%'

'77.07006369426752%'

## Decision Trees

>> Model Making

In [ ]:
from sklearn import tree
DT_model = tree.DecisionTreeClassifier(max_depth=5)
DT_model.fit(inputs_scaled_train,targets_train.iloc[:,1])

DecisionTreeClassifier(max_depth=5)

>> Train accuracy/score

In [ ]:
f'{DT_model.score(inputs_scaled_train,targets_train.iloc[:,1])*100}%'

'81.05975197294251%'

>> Test accuracy/score

In [ ]:
f'{DT_model.score(inputs_scaled_test,targets_test.iloc[:,1])*100}%'

'78.98089171974523%'

## Random Forests

>> Model making

In [ ]:
from sklearn.ensemble import RandomForestClassifier

RF_model = RandomForestClassifier(max_depth = 4,random_state=123,n_estimators=150)
RF_model.fit(inputs_scaled_train,targets_train.iloc[:,1])

RandomForestClassifier(max_depth=4, n_estimators=150, random_state=123)

>> Train accuracy/score

In [ ]:
f'{RF_model.score(inputs_scaled_train,targets_train.iloc[:,1])*100}%'

'83.8782412626832%'

>> Test accuracy/score

In [ ]:
f'{RF_model.score(inputs_scaled_test,targets_test.iloc[:,1])*100}%'

'78.98089171974523%'

In [ ]:
RF_model.predict(inputs_scaled_test),DT_model.predict(inputs_scaled_test)

(array([0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1,
        1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0,
        1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0,
        0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1,
        1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1,
        1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0,
        1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0,
        0, 1, 0], dtype=uint8),
 array([0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1,
        1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0,
        1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0,
        0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0,
        1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1,
        1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0,
      

## KNN

>> Importing

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

>> Data

In [ ]:
knn_train_inputs,knn_train_targets = inputs_scaled_train,targets_train.iloc[:,1]
knn_test_inputs,knn_test_targets = inputs_scaled_test,targets_test.iloc[:,1]

>> Model Making

In [ ]:
knn_model = KNeighborsClassifier(n_neighbors=15,weights='distance')
knn_model.fit(knn_train_inputs,knn_train_targets)

KNeighborsClassifier(n_neighbors=15, weights='distance')

>> Test accuracy/score

In [ ]:
knn_model.score(knn_test_inputs,knn_test_targets)

0.7898089171974523

In [ ]:
knn_train_inputs.shape

(887, 12)

>>Removing Unnecessary Features

In [ ]:
knn_train_inputs = pd.DataFrame(data=knn_train_inputs)
knn_test_inputs = pd.DataFrame(data=knn_test_inputs)
print(knn_model.score(knn_test_inputs,knn_test_targets))
i = 0
while i < knn_train_inputs.shape[1]:
  knn_model_aa = KNeighborsClassifier(n_neighbors=15,weights='distance')
  knn_model_aa.fit(knn_train_inputs.drop(i,axis=1),knn_train_targets)

  knn_model_a = KNeighborsClassifier(n_neighbors=15,weights='distance')
  knn_model_a.fit(knn_train_inputs,knn_train_targets)

  af = knn_model_aa.score(knn_test_inputs.drop(i,axis=1),knn_test_targets)
  bf = knn_model_a.score(knn_test_inputs,knn_test_targets)
  if af > bf:
    knn_train_inputs = knn_train_inputs.drop(i,axis=1)
    knn_test_inputs = knn_test_inputs.drop(i,axis=1)
  i += 1

knn_model = KNeighborsClassifier(n_neighbors=15,weights='distance')
knn_model.fit(knn_train_inputs,knn_train_targets)
print(knn_model.score(knn_test_inputs,knn_test_targets))

0.7898089171974523
0.8280254777070064


>>Finding best K value

In [ ]:
k = 0
accu = 0.
for i in range(3,25,2):
  knn_model = KNeighborsClassifier(n_neighbors=i,weights = 'distance')
  knn_model.fit(knn_train_inputs,knn_train_targets)
  new = knn_model.score(knn_test_inputs,knn_test_targets)
  if new > accu:
    accu = new
    k = i
  print(f'{i} : {new}')
print(f'k : {k}')

3 : 0.7515923566878981
5 : 0.7707006369426752
7 : 0.7834394904458599
9 : 0.8152866242038217
11 : 0.8089171974522293
13 : 0.802547770700637
15 : 0.8280254777070064
17 : 0.8343949044585988
19 : 0.8152866242038217
21 : 0.8152866242038217
23 : 0.821656050955414
k : 17


>>Therefore after doing curse of dimensionality we got an accuracy of 82.80% and now by taking k = 17 we are getting an accuracy of 83.43%

In [ ]:
knn_model = KNeighborsClassifier(n_neighbors=k,weights='distance')
knn_model.fit(knn_train_inputs,knn_train_targets)
print(f'by backelemination and by taking K : {k} we got an accuracy of {knn_model.score(knn_test_inputs,knn_test_targets)*100}%')

by backelemination and by taking K : 17 we got an accuracy of 83.43949044585987%
